rust/projects/file-io/src/kv.rs

```diff
@@ -63,31 +59,137 @@ impl KvStore {
+        mem::swap(&mut self.kv_log, &mut kv_log);
+
+        // Close old log file before removing it. (It's a must on Windows I think)
+        let old_path = kv_log.path.clone();
+        drop(kv_log);
+        fs::remove_file(old_path)?;
+
+        Ok(())
+    }
+}
+
+struct KvLog {
+    path: PathBuf,
+    reader: BufReaderWithPos<File>,
+    writer: BufWriterWithPos<File>,
+    // stores keys and the pos of the last command to modify each
+    index: HashMap<String, CommandPos>,
+    uncompacted: u64,
+}
+
+impl KvLog {
+    // Pay attention that it does not load the log file automatically
+    fn open(path: impl Into<PathBuf>) -> Result<Self> {
+        let path = path.into();
+        let mut writer = BufWriterWithPos::new(
+            OpenOptions::new()
+                .create(true)
+                .read(true)
+                .append(true)
+                .open(&path)?,
+        )?;
+        // Set pos to end of file
+        writer.seek(SeekFrom::End(0))?;
+
+        let reader = BufReaderWithPos::new(File::open(&path)?)?;
+
+        Ok(KvLog {
+            path,
+            reader,
+            writer,
+            index: HashMap::new(),
+            uncompacted: 0,
+        })
     }
 
-    fn load_from_wal(wal_path: impl AsRef<Path>) -> Result<HashMap<String, String>> {
-        let mut map = HashMap::new();
-        let reader = BufReader::new(File::open(wal_path)?);
-        let stream = Deserializer::from_reader(reader).into_iter::<SetCommand>();
-        for set_cmd in stream {
-            let set_cmd = set_cmd?;
-            map.insert(set_cmd.key, set_cmd.value);
+    fn set(&mut self, key: String, value: String) -> Result<()> {
+        let cmd = SetCommand::new(key, value);
+        let pos = self.writer.pos;
+        serde_json::to_writer(&mut self.writer, &cmd)?;
+        self.writer.flush()?;
+        if let Some(old_cmd) = self.index.insert(cmd.key, (pos..self.writer.pos).into()) {
+            self.uncompacted += old_cmd.len;
         }
-        Ok(map)
+        Ok(())
+    }
+
+    fn get(&mut self, key: String) -> Result<Option<String>> {
+        if let Some(cmd_pos) = self.index.get(&key) {
+            self.reader.seek(SeekFrom::Start(cmd_pos.pos))?;
+            let cmd_reader = (&mut self.reader).take(cmd_pos.len);
+            let set_cmd: SetCommand = serde_json::from_reader(cmd_reader)?;
+            Ok(Some(set_cmd.value))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn load(&mut self) -> Result<()> {
+        let mut pos = self.reader.seek(SeekFrom::Start(0))?;
+        let mut stream = Deserializer::from_reader(&mut self.reader).into_iter::<SetCommand>();
+        while let Some(set_cmd) = stream.next() {
+            let new_pos = stream.byte_offset() as u64;
+            if let Some(old_cmd) = self.index.insert(set_cmd?.key, (pos..new_pos).into()) {
+                self.uncompacted += old_cmd.len;
+            }
+            pos = new_pos;
+        }
+        Ok(())
```

Comment:
Since load is a separate step from open, can you put asserts at the top of get and set that verify the index is loaded?

Reply:
I'm hesitating between loading automatically if not loaded and just an assertion at top of get/set. I choose the latter, as the code looks more clear. Being loaded is therefore needed to be ensured elsewhere, but tests should fail easily so mistakes can be easily detected.